equinox-shipment/Domain/FinalizationTransaction.fs

```diff
@@ -0,0 +1,90 @@
+        | FinalizationFailed    of containerId: string
+        interface TypeShape.UnionContract.IUnionContract
+
+    let codec = FsCodec.NewtonsoftJson.Codec.Create<Event>()
+
+type Action =
+    | AssignShipments   of containerId: string * shipmentIds: string[]
+    | FinalizeContainer of containerId: string * shipmentIds: string[]
+    // Reverts the assignment of the container for the shipments provided.
+    | RevertAssignment  of containerId: string * shipmentIds: string[]
+    | Finish of bool
+
+module Fold =
+
+    type State =
+        | Initial
+        | Running   of RunningState
+        | Completed of bool
+    and RunningState =
+        | Assigning of containerId: string * shipmentIds: string[]
+        | Assigned  of containerId: string * shipmentIds: string[]
+        | Reverting of containerId: string * shipmentIds: string[]
+
+    let initial: State = Initial
+
+    let evolve (_: State) (event: Events.Event): State =
+        match event with
+        | Events.FinalizationRequested (containerId, shipmentIds) -> Running (Assigning (containerId, shipmentIds))
+
+        | Events.AssignmentCompleted   (containerId, shipmentIds) -> Running (Assigned  (containerId, shipmentIds))
+        | Events.FinalizationCompleted containerId                -> Completed true
+
+        | Events.RevertRequested       (containerId, shipmentIds) -> Running (Reverting (containerId, shipmentIds))
+        | Events.FinalizationFailed    containerId                -> Completed false
+
+    let nextAction (state: State): Action =
+        match state with
+        | Initial -> failwith "Cannot interpret Initial state"
+        | Running (Assigning (containerId, shipmentIds)) -> Action.AssignShipments   (containerId, shipmentIds)
+        | Running (Assigned  (containerId, shipmentIds)) -> Action.FinalizeContainer (containerId, shipmentIds)
+        | Running (Reverting (containerId, shipmentIds)) -> Action.RevertAssignment  (containerId, shipmentIds)
+        | Completed result                               -> Action.Finish result
+
+    let fold: State -> Events.Event seq -> State =
+        Seq.fold evolve
+
+    let filterValidTransition (event: Events.Event) (state: State) =
+        match state, event with
+        | Initial,               Events.FinalizationRequested _
+        | Running (Assigning _), Events.AssignmentCompleted   _
+        | Running (Assigned _),  Events.FinalizationCompleted _
+        | Running (Assigning _), Events.RevertRequested       _
+        | Running (Reverting _), Events.FinalizationFailed    _ ->
+            Some event
+        | _ -> None
+
+// When there are no event to apply to the state, it pushes the transaction manager to
+// follow up on the next action from where it was.
+let decide (update: Events.Event option) (state : Fold.State) : Action * Events.Event list =
+    let events =
+        update
+        |> Option.bind (fun e -> Fold.filterValidTransition e state)
```

Comment:
can this truly not be expressed more cleanly as a `match` ?

Reply:
Updated to use `match`, thanks!